### PR TITLE
Don't blanket impl LenTextSize

### PR DESCRIPTION
**Pull request by CAD97**
Instead, just provide copy-paste impls for somewhat commonly used std string containers.

### CODE REVIEW

**Comment by CAD97** [comment]
This does mean that `smol_str` either has to impl `LenTextSize` directly, or the user of both `SmolStr` and `text-size` has to use `TextSize::of::<&str>(&SmolStr)`. IIRC, this is only one place in rowan, however, so this shouldn't be a big impact on ra.

File: tests/constructors.rs [hunk 1]
```diff
@@ -1,31 +1,34 @@
...
+#[derive(Copy, Clone)]
+struct BadRope<'a>(&'a [&'a str]);
 
-impl Deref for StringLike<'_> {
-    type Target = str;
-    fn deref(&self) -> &Self::Target {
-        &self.0
+impl LenTextSize for BadRope<'_> {
+    fn len_text_size(self) -> TextSize {
+        self.0.iter().copied().map(LenTextSize::len_text_size).sum()
```
Comment (matklad) [inline tests/constructors.rs]:
```suggestion
        self.0.iter().copied().map(TextSize::of).sum()
```

**Comment by matklad** [comment]
The irony here is that if we made extension trait, and not `TextSize::of`, *the* method to create text sizes, then we'd get deref coercions for free. I do still feel that `::of` design better. I even thing that `of` is better than `of_char` / `of_str`

**Comment by CAD97** [comment]
If only we could rely on the fact that `char: Deref` will never hold, then the following should be enough:

```rust
/// Text-like structures that have a text size.
pub trait TextLen {
    /// The size of this text-alike.
    fn text_len(&self) -> TextSize;
}

impl TextLen for str {
    #[inline]
    fn text_len(&self) -> TextSize {
        self.len().try_into().unwrap()
    }
}

impl TextLen for char {
    #[inline]
    fn text_len(&self) -> TextSize {
        (self.len_utf8() as u32).into()
    }
}

impl<D: Deref> TextLen for D
where
    D::Target: TextLen,
{
    fn text_len(&self) -> TextSize {
        self.deref().text_len()
    }
}
```

With this formulation the error isn't great but at least makes sense (rather than [infinite recursion](https://github.com/rust-lang/rust/issues/47720#issuecomment-604187222)):

```
error[E0277]: the trait bound `(): std::ops::Deref` is not satisfied
  --> tests\constructors.rs:35:26
   |
35 |     let _ = TextSize::of(());
   |                          ^^ the trait `std::ops::Deref` is not implemented for `()`
   | 
  ::: D:\repos\rust-analyzer\text-size\src\size.rs:60:18
   |
60 |     pub fn of<T: TextLen>(text: T) -> TextSize {
   |                  ------- required by this bound in `text_size::size::TextSize::of`
   |
   = note: required because of the requirements on the impl of `text_size::traits::TextLen` for `()`
```

**Pull request merged**